File: 백준/2000/2630/2630.js
// https://www.acmicpc.net/problem/2630

const fs = require('fs');
const stdin = (process.platform === 'linux'
    ? fs.readFileSync('/dev/stdin').toString()
    : `4
    1 1 1 0
    1 1 1 0
    1 1 1 0
    1 1 1 0`
).split('\n');

const input = (() => {
    let line = 0;
    return () => stdin[line++];
})();

const c = +input();
const arr = [];
for (let i = 0; i < c; i++) {
    arr.push(input().trim().split(' ').map(v => +v));
}

let blue = 0;
let white = 0;

const adder = (flag) => {
    if (flag === 1) blue++;
    else white++;
}

const check = (x, y, size) => {
    let flag = arr[x][y];
    for (let i = x; i < x + size; i++) {
        for (let j = y; j < y + size; j++) {
            if (flag !== arr[i][j]) return false;
        }
    }
    return true;
}

const recursive = (x, y, size) => {
    if (size === 0) {
        return;
    }
    const newSize = Math.floor(size / 2);
    if (check(x, y, size)) {
        adder(arr[x][y])
        return;
    }
    else {
        recursive(x, y, newSize);
        recursive(x + newSize, y, newSize);
        recursive(x, y + newSize, newSize);
        recursive(x + newSize, y + newSize, newSize);
    }
}
recursive(0, 0, c);
console.log(white);
console.log(blue);

File: 백준/2000/2630/2630.py
# https://www.acmicpc.net/problem/2630

import sys
count = int(sys.stdin.readline().strip())
arr = [sys.stdin.readline().strip().split() for _ in range(count)]
white = 0
blue = 0


def check(x, y, length):
    flag = arr[x][y]
    for i in range(x, x+length):
        for k in range(y, y+length):
            if flag != arr[i][k]:
                return False
    return True


def recursive(x=0, y=0, length=count):
    global blue, white

    if length == 0:
        return

    if check(x, y, length):
        if arr[x][y] == '1':
            blue += 1
        else:
            white += 1
    else:
        for i in range(x, x+length, length//2):
            for k in range(y, y+length, length//2):
                recursive(i, k, length//2)


recursive()
print(white)
print(blue)
